### DISCUSSION TITLE
Improve rust syntax highlighting.

### DISCUSSION THREAD

**Discussion by gloridifice** [zed-industries/zed#24361] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Some improvement points I found of zed rust syntax highlighting are below.

## lifetime

Color rust lifetime.

Zed ⬇️
<img width="440" alt="Snipaste_2024-02-12_12-14-01" src="https://github.com/zed-industries/zed/assets/39210593/c6f4ef86-82b5-48b8-a11d-93ae0b256ad9">

RustRover (Expected) ⬇️
<img width="180" alt="Snipaste_2024-02-12_12-30-06" src="https://github.com/zed-industries/zed/assets/39210593/fff82882-01df-49a0-a4ca-d3f37db32aa2">
<img width="213" alt="Snipaste_2024-02-12_12-36-37" src="https://github.com/zed-industries/zed/assets/39210593/a7d81c29-8914-4479-96c1-268fe533cf29">

## generic
Color generic with a separate color. (now it's same as "type")
Zed ⬇️
<img width="160" alt="Snipaste_2024-02-12_12-29-20" src="https://github.com/zed-industries/zed/assets/39210593/6cb78339-26f3-4052-a03f-3ac1aae9abc9">
RustRover (Expected) ⬇️
<img width="354" alt="Snipaste_2024-02-12_12-28-25" src="https://github.com/zed-industries/zed/assets/39210593/4ed975a8-a2c0-4f84-bc09-0f1956833094">

## proc macro
Color proc macros with a separate color. (Now it's same as "editor.foreground")
Zed ⬇️
<img width="233" alt="Snipaste_2024-02-12_12-14-20" src="https://github.com/zed-industries/zed/assets/39210593/5cd008b7-89b8-4b22-b136-5fb11db6189f">
<img width="271" alt="Snipaste_2024-02-12_12-14-43" src="https://github.com/zed-industries/zed/assets/39210593/fb4d59c0-e91e-4dfd-b956-d3a4d36f9a4f">

RustRover (Expected) ⬇️
<img width="381" alt="Snipaste_2024-02-12_12-32-57" src="https://github.com/zed-industries/zed/assets/39210593/f8460372-efcd-4481-a1a7-fa393e114d94">
<img width="232" alt="Snipaste_2024-02-12_12-33-04" src="https://github.com/zed-industries/zed/assets/39210593/b0acb701-eba1-4a00-8a2d-5da452a3520d">

## macro's name

Recognize macro name's ! as a whole. (I don't know if this is intentional)

Zed ⬇️
<img width="278" alt="Snipaste_2024-02-12_12-15-24" src="https://github.com/zed-industries/zed/assets/39210593/b1d93531-94a7-4813-a0f7-70ab7eae92cc">

**Comment by universalmind303** [comment]
some additional examples

## zed
<img width="546" alt="image" src="https://github.com/zed-industries/zed/assets/21327470/1890d142-5c42-4838-ac60-9c0dca53154b">

## VSCode
<img width="435" alt="image" src="https://github.com/zed-industries/zed/assets/21327470/5729ca14-ac1d-4c95-bc4a-4b3dce47138b">

**Comment by mascanho** [comment]
Is this still a global issue or is it just my zed? It seems like the syntax highlight lacks a bit, a lot of things share the same colour, especially when compared to neovim. Thank you for the amazing effort everyone.

**Comment by nathansobo** [comment]
@iamnbutler What are your thoughts on coloring lifetimes in our themes? It would indeed be nice to make them stand out a bit more like the examples people are showing.

**Reply by iamnbutler** [reply]
IMO some of the examples are better (like the lifetime ones). The attribute example from rustrover is far worse though, I don't know why you would want to reduce contrast between the attribute and it's values.

**Reply by iamnbutler** [reply]
I'm down with the lifetimes change 👍

**Comment by roycrippen4** [comment]
I really, _really_ like having type arguments/lifetimes both colorized and italicized.

## Neovim
![Image](https://github.com/user-attachments/assets/47ae5b11-305a-4ff4-bf94-f149dbbf38a9)

## Zed
![Image](https://github.com/user-attachments/assets/0a0c9b2e-0d43-4ad2-8b54-d4ac5c931fed)

It would be awesome to reuse my existing tree-sitter queries to merge/override zed's queries and just tweak the colors associated with the matches. Colors that differentiate between primitive types and type identifiers would also be great. 

The lack of a user-facing, fine-grained customization for syntax highlighting is less than ideal. That said, it seems like zed is close to super-customizable and the vim integration probably the best outside of vim/neovim.

**Reply by y1lan** [reply]
I would like to know that which plugin you installed in your neovim to provide such highlight ?

**Comment by dream-dasher** [comment]
What's the current status of this?
The zed documentation currently gives a very short "comprehensive" list of supported syntax elements that can be used in themes. [zed- syntax highlighting](https://zed.dev/docs/extensions/languages#syntax-highlighting).

It seems like a strange weakness of the setup given the TreeSitter based system.

(I was trying to copy a customized Helix theme to Zed for use with Rust and ran into what seem like surprising limitations.)
